pkg/probe/controller.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package probe
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	"istio.io/istio/pkg/log"
+)
+
+var errUnregistered = errors.New("unregistered")
+
+// Controller provides the internal interface to handle events coming
+// from Emitters. Individual controller implementation will update
+// its prober status.
+type Controller interface {
+	io.Closer
+	register(p *Probe, initial error)
+	onChange(p *Probe, newStatus error)
+}
+
+type controllerBase struct {
+	sync.Mutex
+	statuses map[*Probe]error
+	name     string
+}
+
+func (cb *controllerBase) registerBase(p *Probe) {
+	cb.Lock()
+	defer cb.Unlock()
+	if _, ok := cb.statuses[p]; ok {
+		log.Debugf("%s is already registered to %s", p, cb.name)
+		return
+	}
+	cb.statuses[p] = errUnregistered
+}
+
+func (cb *controllerBase) statusLocked() (err error) {
+	if len(cb.statuses) == 0 {
+		return fmt.Errorf("%s has no emitters", cb.name)
+	}
+	for p, status := range cb.statuses {
+		if status != nil {
+			err = multierror.Append(err, fmt.Errorf("%s: %v", p, status))
+		}
+	}
+	return err
+}
+
+func (cb *controllerBase) update(p *Probe, newStatus error) error {
+	cb.Lock()
+	defer cb.Unlock()
+	prev := cb.statusLocked()
+	if stat, ok := cb.statuses[p]; !ok || stat == newStatus {
```

Comment:
So "!ok" means if the prober doesn't exist, we don't update -- why not log an error?